### PR TITLE
[Quant] Qwen-Image text encoder online FP8 quantization

### PR DESCRIPTION
**Pull request by lulugoodcoder**
download (2)

PLEASE FILL IN THE PR DESCRIPTION HERE.
  - vLLM Version: 0.22.0
  - vLLM-Omni Commit: 50be822af995378c1cf3ff8bd5cc743d493363b8
  - Test Result: LPIPS 0.0953 / PSNR 19.82 dB / PASS, plus the 28 + 39 regression passes.
  
## Purpose

  Add online FP8 quantization for the Qwen-Image text encoder (Qwen2.5-VL), mirroring the
  Z-Image text-encoder online-quant pattern (#1338): the text encoder is built on the meta
  device via create_transformers_model so vLLM online quantization (e.g. FP8) is applied to
  its linears, with weights streamed through the loader. The vision tower is dropped on
  meta so its params are never materialized or quantized.

  Enabling this surfaced and fixes three issues that made per-component diffusion
  quantization a silent no-op / crash:

  1. Config propagation — Omni(quantization_config={...}) was dropped on the way to the
  diffusion stage (only diffusion_quantization_config was honored), so per-component dicts
  silently no-op'd (quantized output byte-identical to BF16). Now the generic
  quantization_config kwarg propagates to diffusion stages.
  2. Unmatched linears — ComponentQuantizationConfig returned None for layers outside the
  targeted component (e.g. the DiT transformer's QKVParallelLinear when only text_encoder
  is quantized), which vLLM's LinearBase rejects ("All linear layers should support quant
  method"). It now returns an explicit UnquantizedLinearMethod for unmatched linear layers.
  3. Checkpoint key layout — the Qwen-Image text-encoder checkpoint ships the legacy-flat
  Qwen2.5-VL layout (model.layers.*, visual.*); the instantiated module nests these
  (model.language_model.*, model.visual.*). from_pretrained remaps internally but the
  streaming loader does not, so load_weights now replays the flat→nested remap and skips
  the dropped vision tower.
## Test Plan
 - New quality gate: pytest tests/diffusion/quantization/test_quantization_quality.py -v
  -m "" -k fp8_qwen_image_text_encoder — quantizes only {"text_encoder": {"method":
  "fp8"}}, generates BF16 vs FP8 with the same seed, and fails if LPIPS exceeds 0.20.
  (H100)
  - Regression (CPU): pytest tests/diffusion/quantization/test_fp8_config.py
  tests/diffusion/quantization/test_component_routing.py -v -m "" — covers the shared
  component_config.py / config-routing changes.
  - Manual: python examples/offline_inference/text_to_image/text_to_image.py --model
  Qwen/Qwen-Image (BF16 baseline) and --quantization fp8 (global DiT FP8).
## 

## Test Result
Quality gate — fp8_qwen_image_text_encoder (H100, Qwen/Qwen-Image, 1024×1024, 20 steps,
  seed 142):
  
  - Method: {"text_encoder": {"method": "fp8"}}
  - LPIPS: 0.0953 (threshold 0.20)
  - PSNR: 19.82 dB
  - MAE: 0.043866
  - Result: ✅ PASS  

  Regression suites:

  - test_fp8_config.py — 28 passed
  - test_component_routing.py — 39 passed

  Manual generations on Qwen/Qwen-Image (BF16 baseline and global FP8) look correct.

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by NumberWan** [comment]
A question: this only updates the Qwen-Image, but Qwen-Image-Edit/Layered still load the text encoder via from_pretrained — intentional for this PR, or planned as a follow-up?

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 1]
```diff
@@ -288,38 +296,40 @@ def __init__(
...
+                logger.warning("Qwen-Image: vision tower not found on text encoder; skipping drop")
+
+        self.text_encoder = create_transformers_model(
+            AutoModelForImageTextToText,
+            od_config,
+            hf_config=text_encoder_config,
+            prune_fn=_drop_vision_tower,
+            prefix="text_encoder",
+        ).to(self.device)
+        self.vae = DistributedAutoencoderKLQwenImage.from_pretrained(
```
Comment (NumberWan) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: Small question: VAE loading switched from from_pretrained_with_prefetch to plain from_pretrained. Was that intentional? The old path had a note about subfolder shard races in the old version line 291.

**Comment by NumberWan** [comment]
For text_encoder-only quant, how does DiT init avoid ComponentQuantizationConfig returning None on transformer linears? 

whether get_quant_method should also return UnquantizedLinearMethod when resolve(prefix) is None?

**Comment by lulugoodcoder** [comment]
> For text_encoder-only quant, how does DiT init avoid ComponentQuantizationConfig returning None on transformer linears?
> 
> whether get_quant_method should also return UnquantizedLinearMethod when resolve(prefix) is None?
When a  linear calls get_quant_method:
inside component_config.py
  1. resolve(prefix) finds no component prefix matching (only "text_encoder" is registered) → returns self._default, which is None.
  2. config is None, so the if config is not None: block is skipped.
  3. It falls through to the isinstance(layer, LinearBase) check → returns UnquantizedLinearMethod().

> A question: this only updates the Qwen-Image, but Qwen-Image-Edit/Layered still load the text encoder via from_pretrained — intentional for this PR, or planned as a follow-up?

i can make this change if no one claims it.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py (hunk 1 above)
Comment (lulugoodcoder) [reply vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: thanks for catching this. made an update.

**Comment by lishunyang12** [comment]
Show visual comparision outputs.

**Review by NumberWan — APPROVED** [review]
LGTM

**Comment by lulugoodcoder** [comment]
> Show visual comparision outputs.

uploaded

**Comment by lulugoodcoder** [comment]
@lishunyang12 uploaded the run result

**Comment by lishunyang12** [comment]
Seems like the image quality deteriorated with quant as there is a loss in details? Can you help check it? Can you show the reproducible command you run?

**Comment by lulugoodcoder** [comment]
@lishunyang12 hi, i checked again using below code:
```
#!/usr/bin/env python
"""Generate BF16 baseline vs text-encoder-only FP8 images for Qwen-Image.

The DiT stays BF16; only the text encoder is quantized to FP8 via the
per-component config {"text_encoder": {"method": "fp8"}} — i.e. exactly what
PR #4358 adds. The example CLI can't express per-component quant, so we call the
Omni API directly.

Usage:
    python gen_te_fp8.py                       # defaults (gate settings)
    python gen_te_fp8.py --steps 50 --prompt "a fox in the snow"
"""

import argparse

import torch

from vllm_omni.entrypoints.omni import Omni
from vllm_omni.inputs.data import OmniDiffusionSamplingParams
from vllm_omni.outputs import OmniRequestOutput
from vllm_omni.platforms import current_omni_platform


def generate(omni, prompt, height, width, steps, seed):
    generator = torch.Generator(device=current_omni_platform.device_type).manual_seed(seed)
    outputs = omni.generate(
        {"prompt": prompt},
        OmniDiffusionSamplingParams(
            height=height, width=width, generator=generator, num_inference_steps=steps
        ),
    )
    req_out = OmniRequestOutput.unwrap_result(outputs)
    if not req_out.images:
        raise ValueError("no image in output")
    return req_out.images[0]


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--model", default="Qwen/Qwen-Image")
    p.add_argument("--prompt", default="a cup of coffee on a wooden table, morning light")
    p.add_argument("--seed", type=int, default=142)
    p.add_argument("--steps", type=int, default=20)
    p.add_argument("--height", type=int, default=1024)
    p.add_argument("--width", type=int, default=1024)
    p.add_argument("--outdir", default="out")
    args = p.parse_args()

    import os

    os.makedirs(args.outdir, exist_ok=True)

    # --- BF16 baseline (no quantization) ---
    omni_bf16 = Omni(model=args.model, mode="text-to-image")
    img = generate(omni_bf16, args.prompt, args.height, args.width, args.steps, args.seed)
    img.save(f"{args.outdir}/qwen_bf16.png")
    print(f"saved {args.outdir}/qwen_bf16.png")
    omni_bf16.shutdown()
    del omni_bf16

    # --- Text-encoder-only FP8 (DiT stays BF16) ---
    omni_fp8 = Omni(
        model=args.model,
        mode="text-to-image",
        quantization_config={"text_encoder": {"method": "fp8"}},
    )
    img = generate(omni_fp8, args.prompt, args.height, args.width, args.steps, args.seed)
    img.save(f"{args.outdir}/qwen_te_fp8.png")
    print(f"saved {args.outdir}/qwen_te_fp8.png")
    omni_fp8.shutdown()
    del omni_fp8

    # --- DiT + text-encoder FP8 (flat "fp8" == what --quantization fp8 does on
    #     this branch: DiT quantized selectively per #2795 + text encoder FP8) ---
    omni_both = Omni(
        model=args.model,
        mode="text-to-image",
        quantization_config="fp8",
    )
    img = generate(omni_both, args.prompt, args.height, args.width, args.steps, args.seed)
    img.save(f"{args.outdir}/qwen_dit_te_fp8.png")
    print(f"saved {args.outdir}/qwen_dit_te_fp8.png")
    omni_both.shutdown()


if __name__ == "__main__":
    main()
```
and genetated image as below. dit + te quant did show degrade in quality, however te quant didint. please see pictures attached.

bf16 :
qwen_bf16

dit quant + te quant
qwen_dit_te_fp8

te quant only
qwen_te_fp8

**Review by lishunyang12 — COMMENTED** [review]
meta-device text-encoder streaming reads clean overall -- main thing is you're still dragging the unused `lm_head` through quantization, plus a couple smaller notes on load coverage and a redundant copy.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 2]
```diff
@@ -297,32 +305,42 @@ def __init__(
...
+            # its parameters are never quantized nor allocated on any device.
+            # Handle both transformers layouts: newer puts visual under .model,
+            # older puts it directly on the model.
+            visual_owner = None
+            if hasattr(text_encoder, "model") and hasattr(text_encoder.model, "visual"):
+                visual_owner = text_encoder.model
+            elif hasattr(text_encoder, "visual"):
+                visual_owner = text_encoder
+            if visual_owner is not None:
+                del visual_owner.visual
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py:340]: You drop `visual` here but `lm_head` survives, and the forward only reads `hidden_states[-1]`. So with `{"text_encoder": {"method": "fp8"}}` you allocate, FP8-quantize, and run a full-vocab `lm_head` matmul every encode for logits you throw away. `del` it in this prune fn too.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 3]
```diff
@@ -297,32 +305,42 @@ def __init__(
...
+            else:
+                logger.warning("Qwen-Image: vision tower not found on text encoder; skipping drop")
+
+        self.text_encoder = create_transformers_model(
+            AutoModelForImageTextToText,
+            od_config,
+            hf_config=text_encoder_config,
+            prune_fn=_drop_vision_tower,
+            prefix="text_encoder",
+        ).to(self.device)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: `create_transformers_model` already takes `device=` and materializes params there via `init_parameters`, so this `.to(self.device)` is a second full copy. Pass `device=self.device` and drop the `.to`.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 4]
```diff
@@ -1080,7 +1098,36 @@ def forward(self, req: DiffusionRequestBatch) -> list[DiffusionOutput]:
 
     def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
         loader = AutoWeightsLoader(self)
-        return loader.load_weights(weights)
+        loaded_weights = loader.load_weights(self._remap_text_encoder_weights(weights))
+        # Record components loaded by diffusers submodules to satisfy the
+        # loader's strict coverage check.
+        loaded_weights |= {f"vae.{name}" for name, _ in self.vae.named_parameters()}
+        loaded_weights |= {f"text_encoder.{name}" for name, _ in self.text_encoder.named_parameters()}
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: `text_encoder` now streams through `AutoWeightsLoader`, so this blanket union marks every param loaded regardless, meaning a typo in `_remap_text_encoder_weights` or a genuinely missing weight sails past the strict coverage check silently. Only the quant scale params are ones the loader won't report. Union `named_parameters()` minus what the loader already returned, instead of the whole set.

**Comment by lulugoodcoder** [comment]
@lishunyang12 changed according to your comments.
attached are the pictures generated
qwen_bf16
qwen_bf16 (1)
qwen_dit_te_fp8
qwen_dit_te_fp8 (1)
qwen_te_fp8
qwen_te_fp8 (1)
 from current code.
